fix: Always bind to container selected by user

ad1992 opened the pull request:
fixes https://github.com/excalidraw/excalidraw/issues/5860

- [x] Always bind to container selected by user
- [x] Update the position of bound text to be just after bound text so the z-index is correct
- [x] For text tool keep the logic as its now (Select the container with highest z-index when multiple containers intersect)
- [x] Tests

dwelle commented:
>  Don't bind to container when using text tool as no container is selected. Checked few other tools like Miro, whimsical, Eraser work similar way.

Not so sure about this one. But we shouldn't remove the double-click-on-center-to-bind behavior, which seems to now be gone as well (non-filled element).

ad1992 commented:
> > Don't bind to container when using text tool as no container is selected. Checked few other tools like Miro, whimsical, Eraser work similar way.
> 
> Not so sure about this one. I'd also do that in a separate PR if anything, as it's a (breaking) feature unrelated to this fix.

I am ok with creating a separate PR but the reason I think we should remove is it will have ambiguity which I am fixing in this PR since we don't know which container will be selected hence I find it more intuitive not to bind when using text tool. If we agree I will push that change first and then merge this one. 
> 
> But we shouldn't remove the double-click-on-center-to-bind behavior, which seems to now be gone as well (non-filled element).

Good catch, fixed it and added a spec for the same

dwelle commented on src/element/textWysiwyg.test.tsx:
```diff
@@ -837,25 +871,14 @@ describe("textWysiwyg", () => {
         { id: text.id, type: "text" },
       ]);
 
-      // Attempt to bind another text
-      UI.clickTool("text");
-      mouse.clickAt(
-        rectangle.x + rectangle.width / 2,
-        rectangle.y + rectangle.height / 2,
-      );
-      mouse.down();
-      expect(h.elements.length).toBe(3);
-      text = h.elements[2] as ExcalidrawTextElementWithContainer;
-      editor = document.querySelector(
-        ".excalidraw-textEditorContainer > textarea",
-      ) as HTMLTextAreaElement;
-      await new Promise((r) => setTimeout(r, 0));
-      fireEvent.change(editor, { target: { value: "Whats up?" } });
-      editor.blur();
```
why remove this test?

ad1992 replied:
This test is somewhat incorrect since if a container has bound text then clicking on the center should always edit that text instead of creating a separate text element.
I had spent some time debugging this test, I will add a separate for text tool later. I am not sure why this test was working earlier, probably coz the coords were slightly different when using text tool and since width was too small (1px) so the new text was treated outside container and now since container is being computed earlier so this test started failing.  In any case this test is actually not correct so I modified it.

dwelle replied:
checking it out (with modified assertions), it seems not only a new text is being created, but it's also being bound to the container. It can mean the test conditions are weird / test is not working correctly, or it can mean the implem code incorrectly allows for some edge case that it shouldn't (possibly both, since I can't repro in browser). So I wouldn't totally brush this off, and rather investigate this (put logs into the `startTextEditing` handlers to see why the test is creating a new text, etc.).

dwelle replied:
it's actually due to the initial bound text being marked as deleted (due to `0` dimensions I guess). But even fixing that still breaks it, possibly the hitbox test. So screw it. JSDOM is simply too annoying.

dwelle commented on src/scene/Scene.ts:
```diff
@@ -150,6 +150,19 @@ class Scene {
     this.callbacks.clear();
   }
+
+  insertElementAtIndex(element: ExcalidrawElement, index: number) {
+    const nextElements = [
+      ...this.elements.slice(0, index + 1),
+      element,
+      ...this.elements.slice(index + 1),
+    ];
+    this.replaceAllElements(nextElements);
+  }
```
this doesn't insert at index, but _after_ index.

dwelle commented on src/components/App.tsx:
```diff
@@ -1995,10 +1994,13 @@ class App extends React.Component<AppProps, AppState> {
             isTextElement(selectedElement) ||
             isValidTextContainer(selectedElement)
           ) {
+            const container = selectedElements[0] as ExcalidrawTextContainer;
```
this isn't correct. `selectedElement` can be a text element (= not a container) as per above condition.

ad1992 commented:
@dwelle this PR should be ready to merge

dwelle commented on src/scene/Scene.ts:
```diff
@@ -153,9 +153,9 @@ class Scene {
+      ...this.elements.slice(0, index),
       element,
-      ...this.elements.slice(index + 1),
+      ...this.elements.slice(index),
```
btw, this now means that if ever call this function as you did before the change, passing a result directly from `findIndex`, it will not work. So I'd add a check for `index > -1` and probably throw (so we can discover the bug instead of getting an undefined behavior).

dwelle replied:
dpme/

dwelle commented on src/scene/Scene.ts:
```diff
@@ -150,6 +150,24 @@ class Scene {
   }
+
+  insertElementAtIndex(element: ExcalidrawElement, index: number) {
+    if (Number.isFinite(index) && index < 0) {
```
```suggestion
    if (!Number.isFinite(index) || index < 0) {
```
